SignalServiceKit/src/Messages/OWSReadReceiptManager.m

```diff
@@ -122,51 +145,246 @@ - (void)process
+                                envelope:(OWSSignalServiceProtosEnvelope *)envelope
+{
+    OWSAssert(receiptMessage);
+    OWSAssert(envelope);
+    OWSAssert(receiptMessage.type == OWSSignalServiceProtosReceiptMessageTypeRead);
+
+    if (![self areReadReceiptsEnabled]) {
+        DDLogInfo(@"%@ Ignoring incoming receipt message as read receipts are disabled.", self.tag);
+        return;
+    }
+
+    NSString *recipientId = envelope.source;
+    OWSAssert(recipientId.length > 0);
+
+    PBArray *timestamps = receiptMessage.timestamp;
+
+    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
+        [self.dbConnection readWriteWithBlock:^(YapDatabaseReadWriteTransaction *transaction) {
+            for (int i = 0; i < timestamps.count; i++) {
+                UInt64 timestamp = [timestamps uint64AtIndex:i];
+
+                NSArray<TSOutgoingMessage *> *messages
+                    = (NSArray<TSOutgoingMessage *> *)[TSInteraction interactionsWithTimestamp:timestamp
+                                                                                       ofClass:[TSOutgoingMessage class]
+                                                                               withTransaction:transaction];
+                OWSAssert(messages.count <= 1);
+                if (messages.count > 0) {
+                    // TODO: We might also need to "mark as read by recipient" any older messages
+                    // from us in that thread.  Or maybe this state should hang on the thread?
+                    for (TSOutgoingMessage *message in messages) {
+                        [message updateWithReadRecipient:recipientId transaction:transaction];
+                    }
+                } else {
+                    // Persist the read receipts so that we can apply them to outgoing messages
+                    // that we learn about later through sync messages.
+                    NSString *storageKey = [NSString stringWithFormat:@"%llu", timestamp];
+                    NSSet<NSString *> *recipientIds =
+                        [transaction objectForKey:storageKey inCollection:OWSRecipientReadReceiptCollection];
+                    NSMutableSet<NSString *> *recipientIdsCopy
+                        = (recipientIds ? [recipientIds mutableCopy] : [NSMutableSet new]);
+                    [recipientIdsCopy addObject:recipientId];
+                    [transaction setObject:recipientIdsCopy
+                                    forKey:storageKey
+                              inCollection:OWSRecipientReadReceiptCollection];
+                }
+            }
+        }];
+    });
+}
+
+- (void)outgoingMessageFromLinkedDevice:(TSOutgoingMessage *)message
+                            transaction:(YapDatabaseReadWriteTransaction *)transaction
+{
+    OWSAssert(message);
+    OWSAssert(transaction);
+
+    NSString *storageKey = [NSString stringWithFormat:@"%llu", message.timestamp];
+    NSSet<NSString *> *recipientIds =
```

Comment:
take it or leave it: Rather than "just the data we need right now", I'm generally in favor of storing a wrapper record object, since we often want to add fields, add a little data massaging, or change the structure of the persisted data down the road.

The model class gives us a conventional place to store that migration logic in initWithCoder, otherwise you'll have to do something more complicated later like make a *new* collection at that point, and eagerly migrate stuff from the old collection to the new one, rather than leaving lazy migration in initWithCoder as an option.

So, specifically, I'd recommend a `TSYapDatabaseObject` subclass of type `EarlyReadReceipt` or some such with a "recipientIds" property and the timestamp as the uniqueId.

Reply:
I took a hybrid approach that adds a `TSRecipientReadReceipt` class as you suggest but for now we'll just access its state using class convenience methods.  Should we ever need to add more state, this will provide a easy migration path forward.